scripts/replace-cli-version.sh

```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+VERSIONS=$(gh api -H "Accept: application/vnd.github+json" /repos/github/codeql-cli-binaries/releases | jq -r '.[].tag_name' | head -2)
+
+# we are exporting these variables so that we can access these variables in the workflow
+LATEST_VERSION=$(echo $VERSIONS | awk '{ print $1 }')
+PREVIOUS_VERSION=$(echo $VERSIONS | awk '{ print $2 }')
+
```

Comment:
When we do a minor version bump, we need to add the new version instead of replacing it.

Doing something like this can check if the minor version has changed:

```shell
LATEST_MINOR_VERSION="$(echo $LATEST_VERSION | awk 'BEGIN{FS=".";}{print $1$2}')"
PREVIOUS_MINOR_VERSION="$(echo $PREVIOUS_VERSION | awk 'BEGIN{FS=".";}{print $1$2}')"

if [ $LATEST_MINOR_VERSION != $PREVIOUS_MINOR_VERSION ];
then
   # handle case where minor version is changed
else
  # handle case where patch version is changed
fi
```

Reply:
So the list of versions in `supported_cli_versions.json` is expanding every time we have a minor version bump? For example, if we have the following in the file

```json
[
  "v2.11.5",
  "v2.7.6",
  "v2.8.5",
  "v2.9.4",
  "v2.10.5",
  "nightly"
]
```
and a version `v2.11.6` has been released. Do we want to add that like the following?

```json
[
  "v2.11.6", // this is the new version that got added
  "v2.11.5",
  "v2.7.6",
  "v2.8.5",
  "v2.9.4",
  "v2.10.5",
  "nightly"
]
```

Reply:
`v2.11.5` -> `v2.11.6` would be a patch version bump, so in that case we want to replace!

The minor version bump would be `v2.12.0` in this case, so we want to add it:

```
[
  "v2.12.0", // this is the new minor version that got added
  "v2.11.5",
  "v2.7.6",
  "v2.8.5",
  "v2.9.4",
  "v2.10.5",
  "nightly"
]
```

(Or even, we could add the new minor version _and_ remove the oldest one (i.e. `v.2.7.6`).)

Reply:
I don't think we can actually remove 2.7.6. The removal depends on the support GHES versions and GHES 3.3 is still supported. GHES 3.3 [recommends](https://docs.github.com/en/enterprise-server@3.3/admin/code-security/managing-github-advanced-security-for-your-enterprise/configuring-code-scanning-for-your-appliance#provisioning-the-actions-for-code-scanning) CodeQL 2.7.6, so we can't remove that version yet. In general, I don't think we can remove a version automatically because it's not strictly aligned with the CodeQL CLI releases.

Reply:
GHES 3.3 is only supported for another 6 days (2023-01-01), so it kind of aligns with the CodeQL CLI (give or take 1 week 😄). But we don't have to automate the removal (or at least not in this PR) if we want to keep that separate/manual! 🤔

Reply:
Lol, I always get confused with semver 😂 For some reason I thought number after the second dot was minor release 😅 

I think the removal of a version should be done through a separate workflow/action - from what I can tell, there doesn't seem to be an automated way of telling the which version should be removed. I also think that given a manual trigger with the some inputs (like for example in the image below) and manually adding the version to remove would be better unless we want to completely automate it which may not be as simple or quick 😄 Let me know though whichever way is preferred and I can think on what I can do. PS: I think this issue was for the patch release being released every two weeks or so, so this might benefit from a separate issue which I can take (this PR has been open for a while now - I want to close it 😂 of course if it meets the acceptable standards 😉)
![](https://docs.github.com/assets/cb-93002/images/help/images/workflow-dispatch-inputs.png)

Reply:
Good points @tjgurwara99. I expect that this workflow will always be triggered manually since the timing of when to run this workflow depends on a bunch of human decisions.

I'm personally fine with focusing on patch releases for now and working on a new PR to add minor version bumps. I think we can avoid creating a second workflow for minor version bumps and handle that with a new set of inputs. 

And removing releases when a GHES version drops out of support is probably something that should be done manually since it also means we can remove certain feature flags that only pertain to the older version.

Reply:
I've added the [issue](https://github.com/github/code-scanning/issues/8597) and assigned it to myself 😄 will get to it whenever I get time